### PR TITLE
Fix arena to return aligned memory.

### PR DESCRIPTION
**Pull request by markdroth**
Nico, please make sure I haven't made any dumb mistakes here.  Thanks!

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                 FILE SIZE
 ++++++++++++++ GROWING                   ++++++++++++++
  +0.0%    +110 [None]                    +1.02Ki  +0.0%
   +29%    +114 src/core/lib/gpr/arena.cc    +114   +29%
       +28%     +82 gpr_arena_alloc               +82   +28%
      +143%     +33 gpr_arena_create              +33  +143%

 -------------- SHRINKING                 --------------
  -2.7%     -16 src/core/lib/gpr/alloc.cc     -16  -2.7%
     -13.0%      -9 [Unmapped]                     -9 -13.0%
     -13.7%      -7 gpr_malloc_aligned             -7 -13.7%

  +0.0%    +208 TOTAL                     +1.12Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                 FILE SIZE
 ++++++++++++++ GROWING                   ++++++++++++++
  +0.0%    +110 [None]                    +1.02Ki  +0.0%
   +29%    +114 src/core/lib/gpr/arena.cc    +114   +29%
       +28%     +82 gpr_arena_alloc               +82   +28%
      +143%     +33 gpr_arena_create              +33  +143%

 -------------- SHRINKING                 --------------
  -2.7%     -16 src/core/lib/gpr/alloc.cc     -16  -2.7%
     -13.0%      -9 [Unmapped]                     -9 -13.0%
     -13.7%      -7 gpr_malloc_aligned             -7 -13.7%

  +0.0%    +208 TOTAL                     +1.12Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__16777216_2_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/gpr/alloc.cc [hunk 1]
```diff
@@ -90,8 +90,7 @@ void* gpr_realloc(void* p, size_t size) {
   return p;
 }
 
-void* gpr_malloc_aligned(size_t size, size_t alignment_log) {
-  size_t alignment = ((size_t)1) << alignment_log;
+void* gpr_malloc_aligned(size_t size, size_t alignment) {
   size_t extra = alignment - 1 + sizeof(void*);
```
Comment (nicolasnoble) [inline src/core/lib/gpr/alloc.cc]: Maybe add a `gpr_assert(((alignment - 1) & alignment) == 0)` to make sure we're being passed a power of two ?

Comment (markdroth) [reply src/core/lib/gpr/alloc.cc]: Good idea, done.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                 FILE SIZE
 ++++++++++++++ GROWING                   ++++++++++++++
  +0.0%    +134 [None]                    +1.17Ki  +0.0%
   +29%    +114 src/core/lib/gpr/arena.cc    +114   +29%
       +28%     +82 gpr_arena_alloc               +82   +28%
      +143%     +33 gpr_arena_create              +33  +143%
  +5.5%     +32 src/core/lib/gpr/alloc.cc     +32  +5.5%
       +88%     +45 gpr_malloc_aligned            +45   +88%

  +0.0%    +280 TOTAL                     +1.31Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                 FILE SIZE
 ++++++++++++++ GROWING                   ++++++++++++++
  +0.0%    +134 [None]                    +1.17Ki  +0.0%
   +29%    +114 src/core/lib/gpr/arena.cc    +114   +29%
       +28%     +82 gpr_arena_alloc               +82   +28%
      +143%     +33 gpr_arena_create              +33  +143%
  +5.5%     +32 src/core/lib/gpr/alloc.cc     +32  +5.5%
       +88%     +45 gpr_malloc_aligned            +45   +88%

  +0.0%    +280 TOTAL                     +1.31Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_unary_ping_pong.BM_UnaryPingPong_InProcessCHTTP2_NoOpMutator_NoOpMutator__262144_262144.opt.old: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                cpu_time    real_time
-----------------------------------------------------------------------  ----------  -----------
BM_PumpStreamServerToClient<InProcess>/262144                            +4%         +4%
BM_PumpStreamServerToClient<InProcess>/32768                             +4%         +4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/32768      +9%         +9%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/262144  +4%         +4%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/262144           -7%         -7%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/32768/32768     -5%         -5%
```

**Comment by markdroth** [comment]
Known issues: #13445 #14138 #14139 #13148

**Pull request merged**